[SNAP-1756] Initial changes for Snapshot GII.

rishitesh opened the pull request:
## Changes proposed in this pull request

For transactions with snapshot isolation,  we need to block all write transaction on the regions if there is a GII operation going on. 

To achieve this we use a read-write lock at bucket level. All regions which are doing write operations on a bucket take a read lock. The GII while requesting RVV from target node takes a write lock on the bucket. Hence during GII no write transactions can proceed.

This will ensure when GII starts the region will be in a clean state, either all transactions committed or rolled back. Hence snapshot seen by GII will not be dirty.

 
Please ignore the log statements. That will change before actual merge. 
## Patch testing

Still in progress. 

## ReleaseNotes changes

(Does this change require an entry in ReleaseNotes? If yes, has it been added to it?)

## Other PRs 

(Does this change require changes in other projects- snappydata, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

Commit:
Initial changes for Snapshot GII.

jramnara commented:
What does "Snapshot GII" mean? Is this spec'd ? 

This has to do with making sure GII operations play well with the new Snapshot isolation semantics? and, the idea is to request a snapshot as of when the GII was initiated on the destination server ? and, somehow, the destination would catchup with the in-progress state post the GII ?

rishitesh commented:
@jramnara This PR is not a feature itself. You are right, this PR  is for GII to pick up correct entries from the source region. With snapshot, the actual region might have uncommitted entries. When GII happens, it should ignore those uncommitted entries and picks up correct values from old region entry map.

Commit:
Added a wit for transaction methods in GII.

Commit:
Modified the GII operation to wait for a read-write lock

Commit:
removed unwanted changes.

Commit:
removed unwanted changes.

Commit:
Merge branch 'snappy/master' into SNAP-1756

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/BucketRegion.java:
```diff
@@ -1153,6 +1169,10 @@ private boolean beginLocalWrite(EntryEventImpl event) {
       throw new CacheClosedException("Cache is shutting down");
     }
 
+    if (giiLevelLockForSnapshot) {
```
The write locks are being taken for each put and is being released as the write is getting over. We need to take the lock for the entire duration of the transaction which started because of the write and could be putAlls as well. The read lock should be released in the commit or abort of the transactions.

The write locks which the requestor of gii takes looks good.

Commit:
Added the read lock at bucket level from transactions.
When is region is marked for write in a transaction we take a read lock on the bucket.
Also incase of column tabled we take lock on both row buffer and column table.

Commit:
formatting changes.

Commit:
Added tests for server departure while doing GII.

Commit:
Changed the membership listener to check for only GII requesting member.

Commit:
Cleaned up logs.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/InitialImageOperation.java:
```diff
@@ -4835,6 +4877,53 @@ public static void resetAllGIITestHooks() {
+      // Only release the lock iff requesting member has parted
+      if (this.bucketToMonitor.isHosting() && id.equals(requestingMember)) {
+        BucketRegion br = bucketToMonitor.getHostedBucketRegion();
+        br.releaseSnapshotGIIWriteLock();
```
Can you please check if there is any other case where the gii requestor goes to a new node for gii and ditches the first request for some reason. Because there won't be memberDeparted in that case but we would need to release the write lock on this node nevertheless.

rishitesh replied:
Ok , will check that.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/InitialImageOperation.java:
```diff
@@ -2795,6 +2797,20 @@ final public int getProcessorType() {
+                                     InternalDistributedMember sender) {
+     // Wait for all write operations to get over.
+     BucketRegion bucketRegion = (BucketRegion)rgn;
+      InitializingBucketMembershipObserver listener =
```
Do we need to have a single such listener or clean it up once the gii request is done. Because for every gii request coming to this source for this bucket a new listener will be added. So basically membership listener cleanup.

rishitesh replied:
We are handling only cases where the GII requestor dies down which will be a rare situation. I don't think we should be keeping a listener only for that. Hence cleaning up the listener once GII is either done or requesting member dies.

kneeraj replied:
Exactly. So where are we cleaning the listener registered for each GII request? That was the question actually.

rishitesh replied:
See BucketRegion.releaseSnapshotGIIWriteLock(). The listener is removed while we are releasing the write lock.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -4049,8 +4061,25 @@ public boolean isSnapshot() {
+      if (region instanceof BucketRegion) {
+        BucketRegion br = (BucketRegion)region;
+        BucketRegion companionRegion = br.companionRegion();
+        if (companionRegion != null) {
```
Is companion region for the shadow region self or parent region or null. I am sking becasue there are some optimizations where the cached batch directly land in the shadow region or cached batch region and don't go through row buffer. Is that case covered?

rishitesh replied:
Shadow region's companion region is row buffer & row buffer's companion region is shadow region. We need to take a read lock at one time as GII might lock one and wait for other while transaction has locked the previous one. That will cause a deadlock. 
Direct transfer to cached batch region won't be problematic. Only we will be taking a read lock on row buffer unnecessarily.

kneeraj replied:
The read lock taken on the row buffer might or might not trigger a cached batch creation. But in either case the read lock will be released on commit/rollback only. So I am suggesting to completely not consider the companion region at all. Just the rowbuffer lock is good enough. Isn't it?

kneeraj reviewed (commented):
Gave some comment and some clarification sought. Still reviewing the tests. @suranjan and @swale you also please review it on priority.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/BucketRegion.java:
```diff
@@ -1261,13 +1268,106 @@ public void doUnlockForPrimary() {
+      return companionRegion; // Return null for regions not associated with column tables
+    }
+    synchronized (companionRegionSync) {
+      if (this.getPartitionedRegion().
```
If we are checking for the existence of shadow region or the cached batch region only do we need the 'else' case here. Other type of colocation is not affected by the gii logic put for snapshot. It is only the rowbuffer and the corresponding companion cached batch region. iSn't it?

rishitesh replied:
Correct. As this fix is a workaround and current SI will work for column tables only we can assume this.

kneeraj replied:
So the 'else' part can be completely removed?

kneeraj commented on gemfirexd/tools/src/dunit/java/com/pivotal/gemfirexd/transactions/SnapshotTxGIIDUnit.java:
```diff
@@ -0,0 +1,399 @@
+        System.clearProperty("snappydata.snapshot.isolation.gii.lock");
+      }
+    });
+    super.tearDown2();
```
I think super.tearDown should always be called in try...finally...Here it won't matter I guess as the code here is unlikely to fail but I guess we need that cleanup at any cost hence the suggestion.

rishitesh replied:
Ok..Will change it.

rishitesh replied:
Also, I have plans to add the same tests at Snappy layer once column table insert is transactional.  That would cover the top level APIs.

kneeraj commented on gemfirexd/tools/src/dunit/java/com/pivotal/gemfirexd/transactions/SnapshotTxGIIDUnit.java:
```diff
@@ -0,0 +1,399 @@
+
+    ai2.join();
+
+    RegionVersionVector rvv1 = getRVV(-1);
```
Should we also check whether the rvv is same just after the gii is complete and before ai2 theoretically could have advanced because it would have been waiting on the readlock. so gii and gii with later puts both will be covered.

kneeraj commented on gemfirexd/tools/src/dunit/java/com/pivotal/gemfirexd/transactions/SnapshotTxGIIDUnit.java:
```diff
@@ -0,0 +1,399 @@
+      public void run() {
+        putSomeValues(500);
+      }
+    });
```
Please check the result after the final set of puts.

rishitesh replied:
Sure, Will add the test.  The final set of puts are to ensure , puts does not hang.

kneeraj replied:
Yes. Validating the content will be more reassuring.

kneeraj reviewed (commented):
It will be nice if we add tests from the top level interfaces also like df.... or conn... staements and apis also.

suranjan commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/InitialImageOperation.java:
```diff
@@ -2795,6 +2797,20 @@ final public int getProcessorType() {
+                                     InternalDistributedMember sender) {
+     // Wait for all write operations to get over.
+     BucketRegion bucketRegion = (BucketRegion)rgn;
+      InitializingBucketMembershipObserver listener =
```
Both read and write lock should be taken on both the buckets and this lock taking should be atomic operaion. Otherwise, there could be deadlock. For that we can use only one lock either row or column store bucket.
Other way could be take the lock in same order.

suranjan commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -4049,8 +4061,25 @@ public boolean isSnapshot() {
+      if (region instanceof BucketRegion) {
+        BucketRegion br = (BucketRegion)region;
+        BucketRegion companionRegion = br.companionRegion();
+        if (companionRegion != null) {
```
Both read and write lock should be taken on both the buckets and this lock taking should be atomic operation. Otherwise, there could be deadlock. For that we can use only one lock either row or column store bucket.
Other way could be take the lock in same order.

rishitesh replied:
Hmm. I am changing it to take lock on row buffer. cached batch table will also use row buffer locks both read and write.

rishitesh replied:
@kneeraj In such case it won't be necessary to take companion region.

Commit:
Added more tests.
Incorporated review comments for tests on PR https://github.com/SnappyDataInc/snappy-store/pull/230.

Commit:
Merge branch 'snappy/master' into SNAP-1756

Commit:
Used only one lock for row buffer & cached batch regions.

Commit:
minor correction.

Commit:
Merge branch 'snappy/master' into SNAP-1756

Commit:
Changed the way how a column table row buffer is
identified.
